// @xyo-network/bound-witness/dist/xyo-bound-witness-party.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Monday, 10th December 2018 1:47:35 pm
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-bound-witness-party.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 6th March 2019 1:41:18 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoBoundWitnessParty = void 0;
const base_1 = require("@xyo-network/base");
class XyoBoundWitnessParty extends base_1.XyoBase {
    get keySet() {
        return this.fetter.keySet;
    }
    get signatureSet() {
        return this.witness.signatureSet;
    }
    get heuristics() {
        return this.fetter.heuristics;
    }
    get metadata() {
        return this.witness.metadata;
    }
    constructor(fetter, witness, partyIndex) {
        super();
        this.fetter = fetter;
        this.witness = witness;
        this.partyIndex = partyIndex;
    }
    getHeuristic(schemaObjectId) {
        const items = this.heuristics.filter(h => h.schemaObjectId === schemaObjectId);
        return items.length > 0 ? items[0] : undefined;
    }
    getMetaDataItem(schemaObjectId) {
        const items = this.metadata.filter(m => m.schemaObjectId === schemaObjectId);
        return items.length > 0 ? items[0] : undefined;
    }
}
exports.XyoBoundWitnessParty = XyoBoundWitnessParty;
