// @xyo-network/bound-witness/dist/xyo-keyset.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Monday, 10th December 2018 10:08:31 am
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-keyset.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 1:50:46 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoKeySet = void 0;
const serialization_1 = require("@xyo-network/serialization");
const serialization_schema_1 = require("@xyo-network/serialization-schema");
class XyoKeySet extends serialization_1.XyoBaseSerializable {
    constructor(keys, org) {
        super(serialization_schema_1.schema, org);
        this.keys = keys;
        this.schemaObjectId = serialization_schema_1.schema.keySet.id;
    }
    getData() {
        return this.keys;
    }
    getReadableValue() {
        return this.keys.map(key => key.getReadableValue());
    }
}
exports.XyoKeySet = XyoKeySet;
// tslint:disable-next-line:max-classes-per-file
class XyoKeySetDeserializer {
    constructor() {
        this.schemaObjectId = serialization_schema_1.schema.keySet.id;
    }
    deserialize(data, serializationService) {
        const parseResult = serializationService.parse(data);
        const query = new serialization_1.ParseQuery(parseResult);
        const keys = query.mapChildren(key => serializationService.deserialize(key.readData(true)).hydrate());
        return new XyoKeySet(keys, data);
    }
}
XyoKeySet.deserializer = new XyoKeySetDeserializer();
