// @xyo-network/bound-witness/dist/xyo-signature-set.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Monday, 10th December 2018 10:08:31 am
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-signature-set.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 1:51:09 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoSignatureSet = void 0;
const serialization_1 = require("@xyo-network/serialization");
const serialization_schema_1 = require("@xyo-network/serialization-schema");
class XyoSignatureSet extends serialization_1.XyoBaseSerializable {
    constructor(signatures) {
        super(serialization_schema_1.schema);
        this.signatures = signatures;
        this.schemaObjectId = serialization_schema_1.schema.signatureSet.id;
    }
    getData() {
        return this.signatures;
    }
    getReadableValue() {
        return this.signatures.map(signature => signature.getReadableValue());
    }
}
exports.XyoSignatureSet = XyoSignatureSet;
// tslint:disable-next-line:max-classes-per-file
class XyoSignatureSetDeserializer {
    constructor() {
        this.schemaObjectId = serialization_schema_1.schema.signatureSet.id;
    }
    deserialize(data, serializationService) {
        const parseResult = serializationService.parse(data);
        const query = new serialization_1.ParseQuery(parseResult);
        const signatures = query.mapChildren(sig => serializationService.deserialize(sig.readData(true)).hydrate());
        return new XyoSignatureSet(signatures);
    }
}
XyoSignatureSet.deserializer = new XyoSignatureSetDeserializer();
