// @xyo-network/signing/dist/xyo-stub-public-key.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Thursday, 29th November 2018 4:44:14 pm
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-stub-public-key.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 1:37:53 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoStubPublicKey = void 0;
const serialization_schema_1 = require("@xyo-network/serialization-schema");
const serialization_1 = require("@xyo-network/serialization");
class XyoStubPublicKey extends serialization_1.XyoBaseSerializable {
    static deserialize(data, serializationService) {
        const parsed = serializationService.parse(data);
        return new XyoStubPublicKey(parsed.data.toString('hex'));
    }
    constructor(publicKeyHexString) {
        super(serialization_schema_1.schema);
        this.publicKeyHexString = publicKeyHexString;
        this.schemaObjectId = serialization_schema_1.schema.stubPublicKey.id;
    }
    getRawPublicKey() {
        return Buffer.from(this.publicKeyHexString, 'hex');
    }
    getData() {
        return this.getRawPublicKey();
    }
    getReadableValue() {
        return this.getRawPublicKey().toString('hex');
    }
}
exports.XyoStubPublicKey = XyoStubPublicKey;
XyoStubPublicKey.schemaObjectId = serialization_schema_1.schema.stubPublicKey.id;
