// @xyo-network/serialization/dist/helpers/tree-iterator.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Thursday, 6th December 2018 5:36:11 pm
 * @Email:  developer@xyfindables.com
 * @Filename: tree-iterator.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 1:28:00 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
const base_serializable_1 = require("./base-serializable");
const parse_query_1 = require("./parse-query");
class XyoTreeIterator extends base_serializable_1.XyoBaseSerializable {
    constructor(serializationService, parseResult, readableName, srcBuffer) {
        super(serializationService.schema, srcBuffer);
        this.serializationService = serializationService;
        this.parseResult = parseResult;
        this.readableName = readableName;
    }
    get schemaObjectId() {
        return this.parseResult.id;
    }
    getReadableName() {
        return this.readableName;
    }
    getReadableValue() {
        return Buffer.concat([this.parseResult.headerBytes, this.parseResult.dataBytes]).toString('hex');
    }
    getData() {
        return this.parseResult.dataBytes;
    }
    query() {
        return new parse_query_1.ParseQuery(this.parseResult);
    }
    hydrate() {
        return this.serializationService.hydrate(this.parseResult);
    }
}
exports.XyoTreeIterator = XyoTreeIterator;
