// @xyo-network/network/src/xyo-mock-network-pipe.ts
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Thursday, 29th November 2018 4:23:34 pm
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-mock-network-pipe.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Tuesday, 11th December 2018 9:34:57 am
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */

import { IXyoSerializableObject } from "@xyo-network/serialization"
import { IXyoNetworkPeer, IXyoNetworkPipe } from './@types'

export class XyoMockNetworkPipe implements IXyoNetworkPipe {
  public networkHeuristics: IXyoSerializableObject[] = []

  public sendCount = 0
  public peer: IXyoNetworkPeer = {}
  public initiationData: Buffer | undefined

  constructor (
    private readonly intercepts: Record<string, (data: Buffer, awaitResponse?: boolean) => Promise<Buffer>>,
    private readonly expectedMessages: Buffer[]
  ) { }

  public onPeerDisconnect(callback: (hasError: boolean) => void): () => void {
    console.log(callback)
    return () => undefined
  }

  public async send(data: Buffer, awaitResponse?: boolean): Promise<Buffer> {
    const expected = this.expectedMessages.length > this.sendCount ? this.expectedMessages[ this.sendCount ] : undefined
    if (expected) {
      expect(expected.equals(data)).toBe(true)
    }

    const returnMsg = (
      this.intercepts[ this.sendCount ] && this.intercepts[ this.sendCount ](data, awaitResponse)
    ) || Buffer.alloc(0)

    this.sendCount += 1

    return returnMsg
  }

  public async close(): Promise<void> {
    return
  }
}
