// @xyo-network/network.ble/dist/xyo-client-ble-network.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoClientBluetoothNetwork = void 0;
const base_1 = require("@xyo-network/base");
const logger_1 = require("@xyo-network/logger");
const xyo_nearby_devices_1 = require("./devices/xyo-nearby-devices");
class XyoClientBluetoothNetwork {
    constructor(scanner) {
        this.tryingDevice = false;
        this.nearbyDevices = [];
        this.nearby = new xyo_nearby_devices_1.XyoNearbyDevices();
        this.logger = new logger_1.XyoLogger(false, false);
        this.scanLambda = () => __awaiter(this, void 0, void 0, function* () {
            const nearbyNow = this.scanner.getDevices();
            this.nearbyDevices = this.nearby.nearby(nearbyNow);
            if (!this.tryingDevice && this.nearbyDevices.length > 0) {
                const randomDevice = this.nearbyDevices[Math.floor(Math.random() * this.nearbyDevices.length)];
                this.tryingDevice = true;
                yield this.scanner.stopScan();
                randomDevice.tryCreatePipe().then((createdPipe) => __awaiter(this, void 0, void 0, function* () {
                    if (this.scanInterval) {
                        clearInterval(this.scanInterval);
                    }
                    if (createdPipe) {
                        this.resolveCallback(createdPipe);
                    }
                    else {
                        const callback = this.onClose;
                        if (callback) {
                            callback();
                        }
                        yield this.scanner.startScan();
                        this.tryingDevice = false;
                    }
                })).catch((_e) => __awaiter(this, void 0, void 0, function* () {
                    const callback = this.onClose;
                    this.logger.info(`Error: ${_e}`);
                    if (callback) {
                        callback();
                    }
                    yield this.scanner.startScan();
                    this.tryingDevice = false;
                }));
            }
        });
        this.scanner = scanner;
    }
    resolveCallback(pipe) {
        const callback = this.onPipe;
        if (callback) {
            callback(pipe);
        }
    }
    shutDown() {
        return __awaiter(this, void 0, void 0, function* () {
            const interval = this.scanInterval;
            if (interval) {
                clearInterval(interval);
                this.scanInterval = undefined;
            }
            this.onPipe = undefined;
            this.onClose = undefined;
            yield this.scanner.stopScan();
        });
    }
    waitForJobToFinish() {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, _reject) => {
                this.onClose = () => {
                    resolve();
                };
            });
        });
    }
    findWithTimeout(timeoutInMills) {
        this.tryingDevice = false;
        return new Promise((resolve, _reject) => {
            this.scanner.startScan().then(() => {
                var hasResumed = false;
                const onTimeout = () => __awaiter(this, void 0, void 0, function* () {
                    if (!hasResumed) {
                        if (this.tryingDevice) {
                            yield this.waitForJobToFinish();
                        }
                        yield this.shutDown();
                        resolve(undefined);
                    }
                });
                base_1.XyoBase.timeout(onTimeout, timeoutInMills);
                this.onPipe = (pipe) => __awaiter(this, void 0, void 0, function* () {
                    hasResumed = true;
                    yield this.shutDown();
                    resolve(pipe);
                });
                this.scanInterval = setInterval(this.scanLambda, 1000);
            }).catch((e) => {
                this.logger.info(`Error: ${e}`);
                resolve(undefined);
            });
        });
    }
    find(_catalogue) {
        this.tryingDevice = false;
        return new Promise((resolve, _reject) => __awaiter(this, void 0, void 0, function* () {
            yield this.scanner.startScan();
            this.onPipe = (pipe) => {
                this.shutDown();
                resolve(pipe);
            };
            this.scanInterval = setInterval(this.scanLambda, 1000);
        }));
    }
    stopServer() {
        return this.scanner.stopScan();
    }
}
exports.XyoClientBluetoothNetwork = XyoClientBluetoothNetwork;
