// @xyo-network/network.ble/dist/xyo-client-ble-network.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoClientBluetoothNetwork = void 0;
const base_1 = require("@xyo-network/base");
const logger_1 = require("@xyo-network/logger");
const xyo_nearby_devices_1 = require("./devices/xyo-nearby-devices");
class XyoClientBluetoothNetwork {
    constructor(scanner) {
        this.nearby = new xyo_nearby_devices_1.XyoNearbyDevices();
        this.nearbyDevices = [];
        this.tryingDevice = false;
        this.logger = new logger_1.XyoLogger(false, false);
        this.scanLambda = () => __awaiter(this, void 0, void 0, function* () {
            const nearbyNow = this.scanner.getDevices();
            this.nearbyDevices = this.nearby.nearby(nearbyNow);
            if (!this.tryingDevice && this.nearbyDevices.length > 0) {
                const randomDevice = this.nearbyDevices[Math.floor(Math.random() * this.nearbyDevices.length)];
                this.tryingDevice = true;
                this.logger.info(`XyoClientBluetoothNetwork.scanLambda(): stopScan()`);
                this.scanner.stopScan();
                randomDevice.tryCreatePipe().then((createdPipe) => {
                    if (this.scanInterval)
                        clearInterval(this.scanInterval);
                    if (createdPipe) {
                        this.resolveCallback(createdPipe);
                        this.logger.info(`tryCreatePipe.resolveCallback()`);
                    }
                    else {
                        const callback = this.onClose;
                        if (callback) {
                            return callback();
                        }
                        this.logger.info(`tryCreatePipe.else{ startScan() }`);
                        this.scanner.startScan();
                        this.tryingDevice = false;
                    }
                }).catch((error) => {
                    const callback = this.onClose;
                    if (callback) {
                        return callback();
                    }
                    this.logger.error(`tryCreatePipe.catch(): ${error}`);
                    this.scanner.startScan();
                    this.tryingDevice = false;
                });
            }
        });
        this.scanner = scanner;
    }
    resolveCallback(pipe) {
        const callback = this.onPipe;
        this.logger.info(`resolveCallback()`);
        if (callback)
            callback(pipe); // return pipe / callback(pipe)
    }
    shutDown() {
        const interval = this.scanInterval;
        if (interval) {
            clearInterval(interval);
            this.scanInterval = undefined;
        }
        this.onPipe = undefined;
        this.onClose = undefined;
        this.scanner.stopScan();
        this.logger.info(`XyoClientBluetoothNetwork().shutDown(): stopScan()`);
    }
    waitForJobToFinish() {
        return new Promise((resolve) => {
            this.onClose = () => {
                resolve();
            };
        });
    }
    findWithTimeout(timeoutInMills) {
        this.tryingDevice = false;
        return new Promise((resolve) => {
            this.logger.info(`XyoClientBluetoothNetwork.findWithTimeout()`);
            this.scanner.startScan().then(() => {
                let hasResumed = false;
                const onTimeout = () => __awaiter(this, void 0, void 0, function* () {
                    if (!hasResumed) {
                        if (this.tryingDevice) {
                            this.logger.info(`findWithTimeout().onTimeout: this.tryingDevice { this.waitForJobToFinish }`);
                            this.waitForJobToFinish();
                        }
                        this.logger.info(`findWithTimeout().onTimeout: this.shutDown()`);
                        this.shutDown();
                        resolve(undefined);
                    }
                });
                base_1.XyoBase.timeout(onTimeout, timeoutInMills);
                this.onPipe = (pipe) => __awaiter(this, void 0, void 0, function* () {
                    hasResumed = true;
                    this.shutDown();
                    resolve(pipe);
                });
                this.scanInterval = setInterval(this.scanLambda, 1000);
            }).catch((error) => {
                this.logger.info(`Error: ${error}`);
                resolve(undefined);
            });
        });
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    find(_catalogue) {
        this.tryingDevice = false;
        // console.log(catalogue)
        this.scanner.startScan();
        return new Promise((resolve) => {
            this.logger.info(`XyoClientBluetoothNetwork.find()`);
            this.onPipe = (pipe) => {
                this.shutDown();
                resolve(pipe);
            };
            this.scanInterval = setInterval(this.scanLambda, 1000);
        });
    }
    stopServer() {
        this.logger.info(`XyoClientBluetoothNetwork.stopServer(): stopScan()`);
        return this.scanner.stopScan();
    }
}
exports.XyoClientBluetoothNetwork = XyoClientBluetoothNetwork;
