// @xyo-network/origin-chain/dist/xyo-next-public-key.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Friday, 7th December 2018 12:11:28 pm
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-next-public-key.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 23rd January 2019 12:38:37 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoNextPublicKey = void 0;
const serialization_1 = require("@xyo-network/serialization");
const serialization_schema_1 = require("@xyo-network/serialization-schema");
class XyoNextPublicKey extends serialization_1.XyoBaseSerializable {
    static get schemaObjectId() {
        return XyoNextPublicKey.deserializer.schemaObjectId;
    }
    constructor(publicKey) {
        super(serialization_schema_1.schema);
        this.publicKey = publicKey;
        this.schemaObjectId = serialization_schema_1.schema.nextPublicKey.id;
    }
    getData() {
        return this.publicKey;
    }
    getReadableValue() {
        return this.publicKey.getReadableValue();
    }
}
exports.XyoNextPublicKey = XyoNextPublicKey;
// tslint:disable-next-line:max-classes-per-file
class XyoNextPublicKeyDeserializer {
    constructor() {
        this.schemaObjectId = serialization_schema_1.schema.nextPublicKey.id;
    }
    deserialize(data, serializationService) {
        const parseResult = serializationService.parse(data);
        return new XyoNextPublicKey(serializationService.deserialize(parseResult.dataBytes).hydrate());
    }
}
XyoNextPublicKey.deserializer = new XyoNextPublicKeyDeserializer();
