// @xyo-network/origin-chain/dist/xyo-index.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Friday, 7th December 2018 11:59:00 am
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-index.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 1:56:16 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
const serialization_1 = require("@xyo-network/serialization");
const serialization_schema_1 = require("@xyo-network/serialization-schema");
const buffer_utils_1 = require("@xyo-network/buffer-utils");
class XyoIndex extends serialization_1.XyoBaseSerializable {
    constructor(number, rIndex) {
        super(serialization_schema_1.schema);
        this.number = number;
        this.rIndex = rIndex;
        this.schemaObjectId = serialization_schema_1.schema.index.id;
    }
    getData() {
        if (this.rIndex) {
            return this.rIndex;
        }
        return buffer_utils_1.unsignedIntegerToBuffer(this.number);
    }
    getReadableValue() {
        return this.number;
    }
}
exports.XyoIndex = XyoIndex;
// tslint:disable-next-line:max-classes-per-file
class XyoIndexDeserializer {
    constructor() {
        this.schemaObjectId = serialization_schema_1.schema.index.id;
    }
    deserialize(data, serializationService) {
        const parseResult = serializationService.parse(data);
        const parseQuery = new serialization_1.ParseQuery(parseResult);
        const numberBuffer = parseQuery.readData(false);
        return new XyoIndex(buffer_utils_1.readIntegerFromBuffer(numberBuffer, numberBuffer.length, false, 0), numberBuffer);
    }
}
XyoIndex.deserializer = new XyoIndexDeserializer();
