// @xyo-network/origin-chain/dist/xyo-bridge-block-set.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Monday, 10th December 2018 12:11:55 pm
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-bridge-block-set.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 1:55:26 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
const serialization_1 = require("@xyo-network/serialization");
const serialization_schema_1 = require("@xyo-network/serialization-schema");
class XyoBridgeBlockSet extends serialization_1.XyoBaseSerializable {
    constructor(boundWitnesses) {
        super(serialization_schema_1.schema);
        this.boundWitnesses = boundWitnesses;
        this.schemaObjectId = serialization_schema_1.schema.bridgeBlockSet.id;
    }
    getData() {
        return this.boundWitnesses;
    }
    getReadableValue() {
        return this.boundWitnesses.map(boundWitness => boundWitness.getReadableValue());
    }
}
exports.XyoBridgeBlockSet = XyoBridgeBlockSet;
// tslint:disable-next-line:max-classes-per-file
class XyoBridgeBlockSetDeserializer {
    constructor() {
        this.schemaObjectId = serialization_schema_1.schema.bridgeBlockSet.id;
    }
    deserialize(data, serializationService) {
        const parseResult = serializationService.parse(data);
        const query = new serialization_1.ParseQuery(parseResult);
        const boundWitnesses = query.mapChildren((boundWitness) => {
            return serializationService.deserialize(boundWitness.readData(true)).hydrate();
        });
        return new XyoBridgeBlockSet(boundWitnesses);
    }
}
XyoBridgeBlockSet.deserializer = new XyoBridgeBlockSetDeserializer();
