// @xyo-network/pi-bridge-server/dist/gql/configuration/mutations.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.configurationMutation = void 0;
const withAuth_1 = __importDefault(require("../decorators/withAuth"));
const configurationMutation = () => ({
    setPaymentKey: (0, withAuth_1.default)((parent, { paymentKey }, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        return configuration.setPaymentKey(paymentKey);
    })),
    setDefaultArchivist: (0, withAuth_1.default)((parent, { id }, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        return configuration.setDefaultArchivist(id);
    })),
    attachArchivist: (0, withAuth_1.default)((parent, { dns, port }, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        return configuration.attachArchivist(dns, port);
    })),
    detachArchivist: (0, withAuth_1.default)((parent, args, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        return configuration.detachArchivist(args.id);
    })),
    updatePin: (0, withAuth_1.default)((parent, { oldPin, newPin }, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        return configuration.updatePin(oldPin, newPin);
    })),
    forceUpdate: (0, withAuth_1.default)((parent, args, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        return configuration.forceUpdate();
    })),
    lastResort: (0, withAuth_1.default)((parent, args, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        return configuration.lastResort();
    })),
    checkForUpdates: (0, withAuth_1.default)((parent, args, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        return configuration.checkForUpdates();
    })),
    factoryReset: (0, withAuth_1.default)((parent, args, { configuration }) => __awaiter(void 0, void 0, void 0, function* () {
        yield configuration.factoryReset();
        return true;
    }))
});
exports.configurationMutation = configurationMutation;
