// @xyo-network/pi-bridge-server/dist/gql/configuration/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_tag_1 = __importDefault(require("graphql-tag"));
exports.default = (0, graphql_tag_1.default) `
  type Archivist {
    id: String
    dns: String
    port: Int
  }

  type Version {
    current: String
    latest: String
  }

  extend type Query {
    isConfigured: Boolean
    getAuthToken(pin: String): String
    publicKey: String
    paymentKey: String
    defaultArchivist: Archivist
    archivists: [Archivist]
    getAptPackageVersion: Version
    getNpmPackageVersion: Version
    tailUnattendedUpgrade: String
  }

  extend type Mutation {
    setPaymentKey(paymentKey: String): String
    setDefaultArchivist(id: String): Archivist
    attachArchivist(dns: String, port: Int): Archivist
    detachArchivist(id: String): Archivist
    updatePin(oldPin: String, newPin: String): String
    forceUpdate: Boolean
    lastResort: Boolean
    checkForUpdates: Boolean
    factoryReset: Boolean
  }
`;
