// @xyo-network/hashing/dist/xyo-hash.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Monday, 10th December 2018 9:59:58 am
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-hash.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 1:37:10 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const xyo_base_hash_1 = require("./xyo-base-hash");
const serialization_schema_1 = require("@xyo-network/serialization-schema");
class XyoHash extends xyo_base_hash_1.XyoBaseHash {
    constructor(hash, hashProvider, schemaObjectId) {
        super(serialization_schema_1.schema);
        this.hash = hash;
        this.hashProvider = hashProvider;
        this.schemaObjectId = schemaObjectId;
    }
    getHash() {
        return this.hash;
    }
    verifyHash(data) {
        return __awaiter(this, void 0, void 0, function* () {
            const dataHash = yield this.hashProvider.createHash(data);
            return dataHash.getHash().equals(this.hash);
        });
    }
}
exports.XyoHash = XyoHash;
