// @xyo-network/pi-wifi-ble/dist/characteristics/write/json.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@xyo-network/utils");
const debug_1 = __importDefault(require("debug"));
const create_1 = require("./create");
class WriteJSONCharacteristic extends utils_1.Subscribe {
    constructor(name, uuid) {
        super();
        this.name = name;
        this.uuid = uuid;
        this.logger = (0, debug_1.default)('bleno:write:json');
        this.handleWrite = (body) => {
            try {
                const json = JSON.parse(body);
                this.setCurrentValue(json);
            }
            catch (e) {
                this.logger(this.name, e.message);
            }
        };
        this.characteristic = (0, create_1.createWriteCharacteristic)({ name, uuid }, (body) => this.handleWrite(body), this.logger);
    }
}
exports.default = WriteJSONCharacteristic;
