// @xyo-network/network.ble/dist/devices/xyo-pipe-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoPipeClient = void 0;
const base_1 = require("@xyo-network/base");
const heuristics_common_1 = require("@xyo-network/heuristics-common");
const logger_1 = require("@xyo-network/logger");
const xyo_output_stream_1 = require("../data/xyo-output-stream");
class XyoPipeClient {
    constructor(device) {
        this.networkHeuristics = [];
        this.logger = new logger_1.XyoLogger(false, false);
        this.otherCatalogue = undefined;
        this.initiationData = undefined;
        this.sessionCharacteristic = null;
        this.device = device;
        this.peer = {
            getTemporaryPeerId() {
                return Buffer.from(device.id);
            }
        };
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    onPeerDisconnect(_callback) {
        return () => { this.logger.info(`XyoPipeClient.onPeerDisconnect()`); };
    }
    tryCreatePipe() {
        return __awaiter(this, void 0, void 0, function* () {
            // eslint-disable-next-line no-async-promise-executor
            const timeout = new Promise((_, reject) => __awaiter(this, void 0, void 0, function* () {
                base_1.XyoBase.timeout(() => __awaiter(this, void 0, void 0, function* () {
                    yield this.device.disconnect(); // await?
                    reject(`XyoPipeClient.tryCreatePipe(): timeout()`); // Promise<null>; reject()
                }), 10000);
            }));
            // eslint-disable-next-line no-async-promise-executor
            const promise = new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                try {
                    if (this.device.state !== 'connected') {
                        yield this.device.connect(); // await
                        this.logger.info(`XyoPipeClient.tryCreatePipe(): connect()`);
                    }
                    const services = yield this.device.discoverServicesForUuids(["d684352edf36484ebc982d5398c5593e"]);
                    if (services.length === 1) {
                        const characteristics = yield services[0].discoverCharacteristics();
                        const xyoPipeChar = characteristics.filter((characteristic) => {
                            // characteristic: { uuid: string }
                            this.logger.info(`XyoPipeClient.tryCreatePipe().services.filter()`);
                            return characteristic.uuid === "727a36390eb44525b1bc7fa456490b2d";
                        });
                        if (xyoPipeChar.length === 1) {
                            this.logger.info(`XyoPipeClient.tryCreatePipe().subscribe()`);
                            this.sessionCharacteristic = xyoPipeChar[0];
                            yield xyoPipeChar[0].subscribe();
                            this.networkHeuristics = [heuristics_common_1.rssiSerializationProvider.newInstance(this.device.rssi)];
                            resolve(this);
                            return;
                        }
                        yield this.device.disconnect();
                        reject("No XYO pipe characteristic 1");
                        return;
                    }
                    yield this.device.disconnect();
                    reject("No XYO service");
                }
                catch (error) {
                    // timeout if here
                    this.logger.error(`XyoPipeClient.tryCreatePipe().connect(): ${error}`);
                    yield this.device.disconnect();
                    reject("Timeout error caught");
                }
            }));
            return yield Promise.race([promise, timeout]);
        });
    }
    send(data, awaitResponse) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.info(`XyoPipeClient().send() ${awaitResponse}`);
            if (this.sessionCharacteristic != null) {
                yield this.chunkSend(data, this.sessionCharacteristic);
                if (awaitResponse !== false) {
                    return yield this.read(this.sessionCharacteristic);
                }
                return undefined;
            }
            throw Error("No XYO pipe characteristic 2");
        });
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.info(`XyoPipeClient.close(): disconnect()`);
            yield this.device.disconnect();
        });
    }
    // default no async/await
    read(characteristic) {
        return __awaiter(this, void 0, void 0, function* () {
            // eslint-disable-next-line no-async-promise-executor
            const timeout = new Promise((_, reject) => {
                base_1.XyoBase.timeout(() => {
                    reject(`XyoPipeClient.read(): timeout() ${characteristic.uuid}`);
                }, 12000);
            });
            // eslint-disable-next-line no-async-promise-executor
            const action = new Promise((resolve) => {
                let buffer;
                let bytesReceived = 0;
                const onData = (data, isNotification) => {
                    this.logger.info(`XyoPipeClient().read().onData(): ${isNotification}`);
                    if (isNotification)
                        return;
                    if (bytesReceived === 0) {
                        this.logger.info(`XyoPipeClient().read().onData(): ${bytesReceived}`);
                        bytesReceived = data.readUInt32BE(0);
                        buffer = data.subarray(4, data.length);
                        if (buffer.length === bytesReceived - 4)
                            resolve(buffer);
                        return;
                    }
                    buffer = Buffer.concat([buffer, data]);
                    if (buffer.length === bytesReceived - 4)
                        resolve(buffer);
                };
                characteristic.on("notification", onData);
            });
            return Promise.race([timeout, action]);
        });
    }
    chunkSend(data, characteristic) {
        return __awaiter(this, void 0, void 0, function* () {
            // eslint-disable-next-line no-async-promise-executor
            const timeout = new Promise((_, reject) => __awaiter(this, void 0, void 0, function* () {
                base_1.XyoBase.timeout(() => __awaiter(this, void 0, void 0, function* () {
                    reject("XyoPipeClient.chunkSend(): timeout()");
                }), 10000);
            }));
            // eslint-disable-next-line no-async-promise-executor
            const action = new Promise((resolve) => __awaiter(this, void 0, void 0, function* () {
                const chunksToSend = (0, xyo_output_stream_1.chunkBytes)(this.addBleSize(data), 20);
                this.logger.info(`Sending entire: ${data.toString('hex')}`);
                for (const [, value] of chunksToSend) {
                    yield characteristic.write(chunksToSend[value]);
                }
                // chunksToSend.forEach(async (_, index) => {
                //   await characteristic.write(chunksToSend[ index ])
                // })
                resolve();
            }));
            yield Promise.race([timeout, action]);
        });
    }
    addBleSize(data) {
        const buffer = Buffer.alloc(4);
        buffer.writeUInt32BE(data.length + 4, 0);
        return Buffer.concat([buffer, data]);
    }
}
exports.XyoPipeClient = XyoPipeClient;
