// @xyo-network/pi-wifi-manager/dist/node/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@xyo-network/utils");
const utils_2 = require("../utils");
const lodash_1 = require("lodash");
const debug_1 = __importDefault(require("debug"));
class NodeWifiManager extends utils_1.Subscribe {
    constructor() {
        super();
        this.statusInterval = 10000;
        this.logger = debug_1.default('wifi:node');
        this.wifi = require('node-wifi');
        this.updating = false;
        this.updateStatus = () => __awaiter(this, void 0, void 0, function* () {
            if (this.updating) {
                this.logger('status update in progress');
                return;
            }
            try {
                const status = yield this.getStatus();
                this.setCurrentValue(status);
            }
            catch (e) {
                this.logger('status update error', e.message);
            }
        });
        this.wifi.init({
            iface: null
        });
    }
    getStatus() {
        return __awaiter(this, void 0, void 0, function* () {
            const connections = yield this.wifi.getCurrentConnections();
            const { ssid } = lodash_1.first(connections) || { ssid: '' };
            const ip = utils_2.getIp() || '';
            return {
                ssid,
                ip
            };
        });
    }
    getNetworks() {
        return __awaiter(this, void 0, void 0, function* () {
            const networks = yield this.wifi.getNetworks();
            return networks || [];
        });
    }
    disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.wifi.disconnect();
            return true;
        });
    }
    connect({ ssid, password }) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.wifi.connect({ ssid, password });
        });
    }
    scan() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.wifi.scan();
        });
    }
    subscribe(fn) {
        const unsubscribe = super.subscribe(fn);
        if (this.listenerCount() === 1) {
            this.startStatusListener();
        }
        return () => {
            unsubscribe();
            if (!this.hasSubscribers()) {
                this.stopStatusListener();
            }
        };
    }
    startStatusListener() {
        this.stopStatusListener();
        this.intervalId = setInterval(this.updateStatus, this.statusInterval);
        this.updateStatus();
    }
    stopStatusListener() {
        clearInterval(this.intervalId);
    }
}
exports.default = NodeWifiManager;
