// @xyo-network/serialization/dist/helpers/on-the-fly-serializable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoOnTheFlySerializable = void 0;
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Wednesday, 5th December 2018 11:11:11 am
 * @Email:  developer@xyfindables.com
 * @Filename: on-the-fly-serializable.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 6th March 2019 4:42:51 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
const errors_1 = require("@xyo-network/errors");
const base_serializable_1 = require("./base-serializable");
class XyoOnTheFlySerializable extends base_serializable_1.XyoBaseSerializable {
    constructor(schema, schemaObjectId, dataOptions, readableName, readableValue) {
        super(schema);
        this.schemaObjectId = schemaObjectId;
        this.dataOptions = dataOptions;
        this.readableName = readableName;
        this.readableValue = readableValue;
    }
    getReadableName() {
        return this.readableName;
    }
    getReadableValue() {
        return this.readableValue;
    }
    getData() {
        // tslint:disable-next-line:no-unused-expression
        if (this.dataOptions.buffer) {
            return this.dataOptions.buffer;
        }
        if (this.dataOptions.object) {
            return this.dataOptions.object;
        }
        if (this.dataOptions.array) {
            return this.dataOptions.array;
        }
        if (this.dataOptions.fn) {
            return this.dataOptions.fn();
        }
        throw new errors_1.XyoError(`No option implemented for OnTheFlySerializable`);
    }
}
exports.XyoOnTheFlySerializable = XyoOnTheFlySerializable;
