// @xyo-network/signing.ecdsa/dist/xyo-ecdsa-signature.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Wednesday, 5th December 2018 12:05:59 pm
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-ecdsa-signature.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 1:54:53 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoEcdsaSignature = void 0;
const serialization_1 = require("@xyo-network/serialization");
const serialization_schema_1 = require("@xyo-network/serialization-schema");
class XyoEcdsaSignature extends serialization_1.XyoBaseSerializable {
    constructor(encodedSignature, schemaObjectId, verifyFn) {
        super(serialization_schema_1.schema);
        this.encodedSignature = encodedSignature;
        this.schemaObjectId = schemaObjectId;
        this.verifyFn = verifyFn;
    }
    getData() {
        return this.encodedSignature;
    }
    verify(data, publicKey) {
        return this.verifyFn(this, data, publicKey);
    }
    getReadableValue() {
        return this.encodedSignature.toString('hex');
    }
}
exports.XyoEcdsaSignature = XyoEcdsaSignature;
