// @xyo-network/signing.ecdsa/dist/xyo-ecdsa-secp256k1-uncompressed-public-key.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Tuesday, 20th November 2018 2:11:41 pm
 * @Email:  developer@xyfindables.com
 * @Filename: xyo-ecdsa-secp256k1-uncompressed-public-key.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Wednesday, 12th December 2018 2:28:02 pm
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XyoEcdsaSecp256k1UnCompressedPublicKey = void 0;
const xyo_ecdsa_uncompressed_public_key_1 = require("./xyo-ecdsa-uncompressed-public-key");
const buffer_utils_1 = require("@xyo-network/buffer-utils");
const serialization_schema_1 = require("@xyo-network/serialization-schema");
class XyoEcdsaSecp256k1UnCompressedPublicKey extends xyo_ecdsa_uncompressed_public_key_1.XyoEcdsaUncompressedPublicKey {
    constructor(xCoordinate, yCoordinate, schemaObjectId) {
        super(serialization_schema_1.schema);
        this.xCoordinate = xCoordinate;
        this.yCoordinate = yCoordinate;
        this.schemaObjectId = schemaObjectId;
    }
    getReadableValue() {
        return this.getRawPublicKey().toString('hex');
    }
    get x() {
        return (0, buffer_utils_1.writePointTo32ByteBuffer)(this.xCoordinate);
    }
    get y() {
        return (0, buffer_utils_1.writePointTo32ByteBuffer)(this.yCoordinate);
    }
}
exports.XyoEcdsaSecp256k1UnCompressedPublicKey = XyoEcdsaSecp256k1UnCompressedPublicKey;
