// @xyo-network/network/dist/catalogue-item.js
"use strict";
/*
 * @Author: XY | The Findables Company <ryanxyo>
 * @Date:   Tuesday, 11th December 2018 9:33:01 am
 * @Email:  developer@xyfindables.com
 * @Filename: catalogue-item.ts
 * @Last modified by: ryanxyo
 * @Last modified time: Tuesday, 22nd January 2019 10:19:26 am
 * @License: All Rights Reserved
 * @Copyright: Copyright XY | The Findables Company
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatalogueItem = void 0;
/**
 * The catalogue items this node knows about
 */
var CatalogueItem;
(function (CatalogueItem) {
    CatalogueItem[CatalogueItem["BOUND_WITNESS"] = Math.pow(2, 0)] = "BOUND_WITNESS";
    CatalogueItem[CatalogueItem["TAKE_ORIGIN_CHAIN"] = Math.pow(2, 1)] = "TAKE_ORIGIN_CHAIN";
    CatalogueItem[CatalogueItem["GIVE_ORIGIN_CHAIN"] = Math.pow(2, 2)] = "GIVE_ORIGIN_CHAIN";
    CatalogueItem[CatalogueItem["TAKE_REQUESTED_BLOCKS"] = Math.pow(2, 3)] = "TAKE_REQUESTED_BLOCKS";
    CatalogueItem[CatalogueItem["GIVE_REQUESTED_BLOCKS"] = Math.pow(2, 4)] = "GIVE_REQUESTED_BLOCKS";
})(CatalogueItem || (exports.CatalogueItem = CatalogueItem = {}));
