// @xyo-network/ble-central.noble/dist/noble-descriptor.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NobleDescriptor = void 0;
const logger_1 = require("@xyo-network/logger");
class NobleDescriptor {
    constructor(descriptor) {
        this.logger = new logger_1.XyoLogger(false, false);
        this.descriptor = descriptor;
    }
    get uuid() {
        return this.descriptor.uuid;
    }
    get name() {
        return this.descriptor.name;
    }
    get type() {
        return this.descriptor.type;
    }
    readValue() {
        // eslint-disable-next-line no-async-promise-executor
        return new Promise((resolve, reject) => {
            this.logger.info(`Trying to read value on descriptor with uuid: ${this.descriptor.uuid}`);
            try {
                this.logger.info(`Successfully read value of descriptor with uuid: ${this.descriptor.uuid}`);
                this.descriptor.readValue((_, data) => {
                    resolve(data);
                });
            }
            catch (e) {
                this.logger.error(`Error reading value of descriptor with uuid: ${this.descriptor.uuid}`);
                reject(`readValue(): Error ${e}`);
            }
        });
    }
    writeValue(value) {
        // eslint-disable-next-line no-async-promise-executor
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            this.logger.info(`Trying to write value to descriptor with uuid: ${this.descriptor.uuid}`);
            try {
                this.logger.info(`Successfully wrote value of descriptor with uuid: ${this.descriptor.uuid}`);
                this.descriptor.writeValue(value);
                resolve();
            }
            catch (e) {
                this.logger.error(`Error writing value of descriptor with uuid: ${this.descriptor.uuid}`);
                reject(`writeValue(): Error ${e}`);
            }
        }));
    }
}
exports.NobleDescriptor = NobleDescriptor;
