// @xyo-network/ble-central.noble/dist/noble-scan.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NobleScan = void 0;
const logger_1 = require("@xyo-network/logger");
// import noble from '@s524797336/noble-mac'
const noble_1 = __importDefault(require("@abandonware/noble"));
const noble_device_1 = require("./noble-device");
class NobleScan {
    constructor() {
        this.logger = new logger_1.XyoLogger(false, false);
        this.inRangeDevices = {};
        this.cleanDevices = () => {
            for (const id in this.inRangeDevices) {
                if (Date.now() - (this.inRangeDevices[id].lastSeen) > 5000) {
                    delete this.inRangeDevices[id];
                }
            }
        };
        this.stateChange = (state) => {
            this.logger.info(`Noble state change: ${state}`);
        };
        this.discover = (peripheral) => {
            const entered = !this.inRangeDevices[peripheral.id];
            if (entered) {
                this.inRangeDevices[peripheral.id] = new noble_device_1.NobleDevice(peripheral);
            }
            this.inRangeDevices[peripheral.id].lastSeen = Date.now();
        };
        noble_1.default.on("stateChange", this.stateChange);
        noble_1.default.on("discover", this.discover);
        setInterval(this.cleanDevices, 1000);
    }
    startScan() {
        return new Promise((resolve, reject) => {
            if (noble_1.default.state === "poweredOn") {
                const callback = () => {
                    this.logger.info("Scanner stared successfully");
                    noble_1.default.removeListener('scanStart', undefined);
                    resolve();
                };
                this.logger.info("Trying to start scanner");
                noble_1.default.on('scanStart', callback);
                noble_1.default.startScanning([], false);
                return;
            }
            this.logger.error("Noble is not powered on");
            reject();
        });
    }
    stopScan() {
        return new Promise((resolve) => {
            this.logger.info("Stopping noble scanner");
            noble_1.default.stopScanning();
            noble_1.default.removeListener('scanStop', undefined);
            resolve();
        });
    }
    getDevices() {
        const returnArray = [];
        for (const [, value] of Object.entries(this.inRangeDevices)) {
            returnArray.push(value);
        }
        return returnArray;
    }
}
exports.NobleScan = NobleScan;
